[AutoFill Debugging] Automatically extract items from same-origin subframes

whsieh opened the pull request:
#### 256ca142887492fec4008131e298b20ff29189de
<pre>
[AutoFill Debugging] Automatically extract items from same-origin subframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=304789">https://bugs.webkit.org/show_bug.cgi?id=304789</a>
<a href="https://rdar.apple.com/167349874">rdar://167349874</a>

Reviewed by Abrar Rahman Protyasha.

Work towards support for subframes in text extraction output by automatically recursing into only
same-origin subframes. See below for more details.

Test: http/tests/text-extraction/debug-text-extraction-subframes.html

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/text-extraction/debug-text-extraction-subframes-expected.txt: Added.
* LayoutTests/http/tests/text-extraction/debug-text-extraction-subframes.html: Added.

Add a new layout test to verify that we recurse into nested same-origin subframes, but avoid
extracting items from nested cross-origin subframes.

* LayoutTests/http/tests/text-extraction/resources/subframe1.html: Added.
* LayoutTests/http/tests/text-extraction/resources/subframe2.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractItemData):
(WebCore::TextExtraction::areSameOrigin):
(WebCore::TextExtraction::extractRecursive):

When we hit a same-origin `iframe` element with a local content frame, use `extractItem` below to
recurse into the frame, and add all its child items as children of the `iframe` item.

(WebCore::TextExtraction::makeRootItem):
(WebCore::TextExtraction::extractItem):

Split this out into a separate method that takes a `LocalFrame` instead of `Page`.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:

Add a new data type, `IFrameData`, that represents an extracted `iframe` element; we currently only
include the security origin as a string in the output, and also include a `FrameIdentifier` (which
we&apos;ll use in a subsequent patch).

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::addPartsForItem):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::createItemWithChildren):

Canonical link: <a href="https://commits.webkit.org/305055@main">https://commits.webkit.org/305055@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2da2a3402c6a4398fa3270fdf592514d6912fcb6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137329 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9689 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48616 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/145080 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90302 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d0567ce4-e487-4059-a655-177687624aed) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139201 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10393 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9816 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/105010 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f0353932-ff88-458b-b3cd-1da9a1c7f280) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140274 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7679 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123076 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85866 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1a187b7b-ca4a-4a46-903f-1ead5646ccf3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7316 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5034 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5667 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116680 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41229 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147837 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9372 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41789 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113382 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9390 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7888 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113723 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28875 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7236 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119319 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63954 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9421 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37373 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9151 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72986 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9361 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9213 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
Change LGTM, two testing suggestions, do you want to add these tests?

1. cross-origin iframe (that verifies content isn't leaked)
2. nested (same-origin) iframe

whsieh committed:
[AutoFill Debugging] Automatically extract items from same-origin subframes
https://bugs.webkit.org/show_bug.cgi?id=304789
rdar://167349874

Reviewed by Abrar Rahman Protyasha.

Work towards support for subframes in text extraction output by automatically recursing into only
same-origin subframes. See below for more details.

Test: http/tests/text-extraction/debug-text-extraction-subframes.html

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/text-extraction/debug-text-extraction-subframes-expected.txt: Added.
* LayoutTests/http/tests/text-extraction/debug-text-extraction-subframes.html: Added.

Add a new layout test to verify that we recurse into nested same-origin subframes, but avoid
extracting items from nested cross-origin subframes.

* LayoutTests/http/tests/text-extraction/resources/subframe1.html: Added.
* LayoutTests/http/tests/text-extraction/resources/subframe2.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractItemData):
(WebCore::TextExtraction::areSameOrigin):
(WebCore::TextExtraction::extractRecursive):

When we hit a same-origin `iframe` element with a local content frame, use `extractItem` below to
recurse into the frame, and add all its child items as children of the `iframe` item.

(WebCore::TextExtraction::makeRootItem):
(WebCore::TextExtraction::extractItem):

Split this out into a separate method that takes a `LocalFrame` instead of `Page`.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:

Add a new data type, `IFrameData`, that represents an extracted `iframe` element; we currently only
include the security origin as a string in the output, and also include a `FrameIdentifier` (which
we'll use in a subsequent patch).

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::addPartsForItem):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::createItemWithChildren):

Canonical link: https://commits.webkit.org/305055@main
